Core/PersistentPixelStoring.swift

```diff
@@ -0,0 +1,157 @@
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import Networking
+
+public struct PersistentPixelMetadata: Codable, Equatable {
+    enum PixelType: Codable {
+        case daily
+        case count
+        case regular
+    }
+
+    let eventName: String
+    let pixelType: PixelType
+    let additionalParameters: [String: String]
+    let includedParameters: [Pixel.QueryParameters]
+
+    var pixelName: String {
+        switch pixelType {
+        case .daily: return eventName + "_d"
+        case .count: return eventName + "_c"
+        case .regular: return eventName
+        }
+    }
+
+    var timestamp: String? {
+        return additionalParameters[PixelParameters.originalPixelTimestamp]
+    }
+}
+
+protocol PersistentPixelStoring {
+    func append(pixel: PersistentPixelMetadata) throws
+    func replaceStoredPixels(with pixels: [PersistentPixelMetadata]) throws
+    func storedPixels() throws -> [PersistentPixelMetadata]
+}
+
+public enum PersistentPixelStorageError: Error {
+    case readError(Error)
+    case writeError(Error)
+    case encodingError(Error)
+    case decodingError(Error)
+}
+
+final class DefaultPersistentPixelStorage: PersistentPixelStoring {
+
+    enum Constants {
+        static let queuedPixelsFileName = "queued-pixels.json"
+        static let pixelCountLimit = 100
+    }
+
+    private let fileManager: FileManager
+    private let fileName: String
+    private let storageDirectory: URL
+    private let pixelCountLimit: Int
+
+    private let fileAccessQueue = DispatchQueue(label: "Persistent Pixel File Access Queue", qos: .utility)
+    private let encoder = JSONEncoder()
+    private let decoder = JSONDecoder()
+
+    private var fileURL: URL {
+        return storageDirectory.appendingPathComponent(fileName)
+    }
+
+    init(fileManager: FileManager = .default,
+         fileName: String = Constants.queuedPixelsFileName,
+         storageDirectory: URL? = nil,
+         pixelCountLimit: Int = Constants.pixelCountLimit) {
+        self.fileManager = fileManager
+        self.fileName = fileName
+        self.pixelCountLimit = pixelCountLimit
+
+        if let storageDirectory = storageDirectory {
+            self.storageDirectory = storageDirectory
+        } else if let appSupportDirectory = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first {
+            self.storageDirectory = appSupportDirectory
+        } else {
+            fatalError("Unable to locate application support directory")
+        }
+    }
+
+    func append(pixel: PersistentPixelMetadata) throws {
+        try fileAccessQueue.sync {
+            var pixels = try self.readStoredPixelDataFromFileSystem()
+            pixels.append(pixel)
+
+            if pixels.count > pixelCountLimit {
+                pixels.removeFirst()
```

Comment:
That drop does not necessairly remove the oldest - because of how the failed pixel array is behaving.

Reply:
The order of the array is now preserved - pixels get written immediately upon failure, and we filter them out when they get successfully retried, so we're no longer juggling two arrays at once and combining them in unexpected ways.

That said, it's still possible for pixels to get a little bit out of order; for example:
1. The app tries to send pixel 1, and then tries to send pixel 2 a second later
2. Pixel 1, for whatever reason, takes longer for its network request to fail
3. Because we capture the timestamp at the time of the pixel send initiation, pixel 1 is now positioned _after_ pixel 2 in the storage despite having a timestamp 1 second earlier

This feels low risk to me so I've opted to leave it, but adding a sorting mechanism to the storage call wouldn't be terribly expensive; let me know how you feel.